File: go/pascals-triangle/pascals_triangle.go
// Package pascal computes Pascal's Triangle.
package pascal

import (
	"errors"
)

// Triangle returns Pascal's Triangle for a given number.
func Triangle(limit int) [][]int {
	if limit <= 0 {
		return [][]int{}
	}

	triangle := [][]int{}

	for n := 0; n < limit; n++ {
		line := []int{1}

		for k := 1; k <= n; k++ {
			c, _ := NChooseK(n, k)
			line = append(line, c)
		}

		triangle = append(triangle, line)
	}

	return triangle
}

// Factorial returns the factorial of a number.
// https://en.wikipedia.org/wiki/Factorial
func Factorial(n int) (int, error) {
	if n < 0 {
		return 0, errors.New("Factorial(n): n must be >= 0")
	}

	f := 1

	for i := 2; i <= n; i++ {
		f *= i
	}

	return f, nil
}

// NChooseK returns the number of possible ways to choose 2 numbers from a set.
// https://en.wikipedia.org/wiki/Binomial_coefficient
func NChooseK(n, k int) (int, error) {
	if k < 0 {
		return 0, errors.New("NChooseK(n, k): k must be >= 0")
	}

	if n < k {
		return 0, errors.New("NChooseK(n, k): n must be >= k")
	}

	// Note: The fast return checks above prevent these from returning errors.
	t1, _ := Factorial(n)
	t2, _ := Factorial(k)
	t3, _ := Factorial(n - k)

	return t1 / (t2 * t3), nil
}

File: cpp/pascals-triangle/pascals_triangle.cpp
#include "pascals_triangle.hpp"

namespace pascals_triangle {

triangle_t generate_rows(int limit) {
    triangle_t triangle{};

    if (limit <= 0) {
        return triangle;
    }

    // NOLINTNEXTLINE(readability-identifier-length)
    for (int n = 0; n < limit; n++) {
        row_t row{1};

        for (int k = 1; k <= n; k++) {
            const int coefficient = n_choose_k(n, k);
            row.emplace_back(coefficient);
        }

        triangle.emplace_back(row);
    }

    return triangle;
}

/* returns the factorial of a number.
 * https://en.wikipedia.org/wiki/Factorial
 */
int factorial(int number) {
    if (number < 0) {
        return -1;
    }

    int result{1};

    for (int i = 2; i <= number; i++) {
        result *= i;
    }

    return result;
}

/* returns the number of possible ways to choose 2 numbers from a set.
 * https://en.wikipedia.org/wiki/Binomial_coefficient
 */
// NOLINTNEXTLINE(readability-identifier-length,bugprone-easily-swappable-parameters)
int n_choose_k(int n, int k) {
    if (k < 0 or n < k) {
        return -1;
    }

    int result{};

    const int factorial1 = factorial(n);
    const int factorial2 = factorial(k);
    const int factorial3 = factorial(n - k);

    result = factorial1 / (factorial2 * factorial3);

    return result;
}

} // namespace pascals_triangle
